[Merged by Bors] - feat(data/list/*): add indexed versions of some list functions

JLimperg opened the pull request:
Add `foldr_with_index`, `foldl_with_index`, `mfoldr_with_index`, `mfoldl_with_index`, `mmap_with_index` and `mmap_with_index'`. The new functions are proven correct by relating them to their non-indexed counterparts.

JLimperg committed:
feat(data/list/defs): Add list.*_with_index

Add `foldr_with_index`, `foldl_with_index`, `mmap_with_index` etc.

JLimperg committed:
Add lemmas about foldr_with_index, foldl_with_index etc.

JLimperg committed:
Add lemmas about mfold*_with_index and mmap_with_index[']

JLimperg committed:
Make all the lemmas into theorems

JLimperg committed:
Remove @[simp] from helper lemmas

JLimperg committed:
Placate the linter

kim-em commented on src/data/list/basic.lean:
```diff
@@ -1304,6 +1305,10 @@ end insert_nth
 @[simp] lemma map_nil (f : α → β) : map f [] = [] := rfl
 
+theorem map_eq_foldr (f : α → β) (l : list α)
+  : map f l = foldr (λ a bs, f a :: bs) [] l :=
```
The style guide says not to start lines with punctuation. 
```suggestion
theorem map_eq_foldr (f : α → β) (l : list α) :
  map f l = foldr (λ a bs, f a :: bs) [] l :=
```

kim-em replied:
And below.

JLimperg replied:
Fixed throughout.

kim-em commented on src/data/list/basic.lean:
```diff
@@ -1363,6 +1368,10 @@ begin
 
+theorem map_filter_eq_foldr (f : α → β) (p : α → Prop) [decidable_pred p] (as : list α)
+  : map f (filter p as) = foldr (λ a bs, if p a then f a :: bs else bs) [] as :=
+by induction as; [ refl, simp! [*, apply_ite (map f)] ]
```
Please don't use `X; [ Y, Z ]` when you could just write `{ X, { Y }, { Z } }`. It's unnecessarily obfuscatory.

JLimperg replied:
Fixed throughout.

kim-em commented on src/data/list/basic.lean:
```diff
@@ -1304,6 +1305,10 @@ end insert_nth
 @[simp] lemma map_nil (f : α → β) : map f [] = [] := rfl
 
+theorem map_eq_foldr (f : α → β) (l : list α)
+  : map f l = foldr (λ a bs, f a :: bs) [] l :=
```
Might as well use dot notation when possible? Also below.

JLimperg replied:
I dislike dot notation and therefore generally avoid it unless it allows me to omit a namespace. But I can also dotify everything if you find this clearer.

kim-em commented on src/data/list/defs.lean:
```diff
@@ -170,13 +170,88 @@ def find (p : α → Prop) [decidable_pred p] : list α → option α
+
+/-- `indexes l` is the list of indexes of `l`, i.e. `[0 .. l.length - 1]`. -/
+def indexes (l : list α) : list ℕ :=
+l.foldr_with_index (λ i _ is, i :: is) []
```
Why even introduce this? It seems easier for people just to use `list.range (l.length)`, rather than giving an obscure construction of this list and then have to prove that it is the same.

JLimperg replied:
Removed.

kim-em commented on src/data/list/defs.lean:
```diff
@@ -170,13 +170,88 @@ def find (p : α → Prop) [decidable_pred p] : list α → option α
+
+     indexed l = zip (indexes l) l -/
+def indexed (l : list α) : list (ℕ × α) :=
+l.foldr_with_index (λ i a out, (i, a) :: out) []
```
We already have this as [list.enum](https://github.com/leanprover-community/lean/blob/f539be1/library/init/data/list/basic.lean#L268), so please remove.

JLimperg replied:
Replaced with `list.enum`.

kim-em commented on src/data/list/defs.lean:
```diff
@@ -170,13 +170,88 @@ def find (p : α → Prop) [decidable_pred p] : list α → option α
+
+/-- `indexes_of a l` is the list of all indexes of `a` in `l`.
+
+     indexes_of a [a, b, a, a] = [0, 2, 3] -/
```
Add backticks, and maybe say something like "For example".

kim-em replied:
(I realise these problems predate your PR!)

bryangingechen replied:
Backticks shouldn't be necessary since the indentation already causes these to be rendered as code blocks in VS Code and on the doc page.

JLimperg replied:
What Bryan says. But Scott is now the second maintainer to get confused by this, so I've changed it anyway.

kim-em commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,168 @@
+  { simp [indexes, foldr_with_index_correct'], exact h },
+  rw ←foldr_map,
+  simp [map_fst_zip]
+end
```
Hopefully remove, if `indexes` is removed?

JLimperg replied:
Removed.

kim-em commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,168 @@
+/-- Specification of `foldr_with_index_aux`. -/
+def foldr_with_index_aux_spec (f : ℕ → α → β → β) (start : ℕ) (b : β)
+  (as : list α) : β :=
+foldr (uncurry f) b $ zip (range' start as.length) as
```
This isn't the specification, it's an alternative implementation. Maybe I'm confused by what's going on here.

JLimperg replied:
Well, one way to specify a function is to say that it behaves like another (less efficient) function. That's my strategy for all these functions because I don't want to redo all the `foldr` lemmas for `foldr_with_index`.

kim-em commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,168 @@
+  = f start a (foldr_with_index_aux_spec f (start + 1) b as) := rfl
+
+theorem foldr_with_index_aux_correct (f : ℕ → α → β → β) (start b as)
+  : foldr_with_index_aux f start b as = foldr_with_index_aux_spec f start b as :=
```
We don't usually use this `_correct` naming scheme.

kim-em replied:
Just say more in the name, e.g. along the lines of `X_eq_Y`.

JLimperg replied:
Fixed throughout.

kim-em commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,168 @@
+  : foldr_with_index f b as = foldr (uncurry f) b (zip (range as.length) as) :=
+by simp only
+    [ foldr_with_index, foldr_with_index_aux_spec, foldr_with_index_aux_correct
+    , ←range_eq_range' ]
```
Punctuation rule again.

JLimperg replied:
Fixed throughout.

JLimperg committed:
reformatting

JLimperg committed:
Avoid t1; [ t2, t3 ] tactic notation

JLimperg committed:
backticks instead of indentation

JLimperg committed:
Remove list.{indexes,indexed}

list.indexed is replaced by list.enum.

JLimperg committed:
Rename *_correct lemmas

JLimperg committed:
Remove unused imports

Vierkantor committed:
Fix markdown issues

Apparently, Github has problemns with parsing lines starting with ```, so let's see if this fixes it.

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+  foldr_with_index f b as = foldr (uncurry f) b (enum as) :=
+by simp only
+    [ foldr_with_index, foldr_with_index_aux_spec,
+      foldr_with_index_aux_eq_foldr_with_index_aux_spec, enum ]
```
```suggestion
    [foldr_with_index, foldr_with_index_aux_spec,
     foldr_with_index_aux_eq_foldr_with_index_aux_spec, enum]
```

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+  find_indexes p as = map prod.fst (indexes_values p as) :=
+by simp only
+    [ indexes_values_eq_filter_enum, map_filter_eq_foldr, find_indexes,
+      foldr_with_index_eq_foldr_enum, uncurry ]
```
```suggestion
    [indexes_values_eq_filter_enum, map_filter_eq_foldr, find_indexes,
     foldr_with_index_eq_foldr_enum, uncurry]
```

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+  foldl (λ a (p : ℕ × β), f p.fst a p.snd) a (enum bs) :=
+by simp only
+    [ foldl_with_index, foldl_with_index_aux_spec,
+      foldl_with_index_aux_eq_foldl_with_index_aux_spec, enum ]
```
```suggestion
    [foldl_with_index, foldl_with_index_aux_spec,
     foldl_with_index_aux_eq_foldl_with_index_aux_spec, enum]
```

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+theorem mfoldl_with_index_eq_mfoldl_enum [is_lawful_monad m] {α β}
+  (f : ℕ → β → α → m β) (b : β) (as : list α) :
+  mfoldl_with_index f b as =
+  mfoldl (λ b (p : ℕ × α), f p.fst b p.snd) b (enum as) :=
```
```suggestion
    mfoldl (λ b (p : ℕ × α), f p.fst b p.snd) b (enum as) :=
```

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+    b ← f start a,
+    bs ← mmap_with_index_aux_spec f (start + 1) as,
+    pure $ b :: bs
+  } :=
```
```suggestion
    do b ← f start a,
      bs ← mmap_with_index_aux_spec f (start + 1) as,
      pure $ b :: bs :=
```

You could also rewrite this in terms of `applicative`, but I guess that doesn't make it easier to work with?

JLimperg replied:
No you're right, the spec should use `traverse` instead of the unnecessarily constrained `mmap`. Let me fix that...

JLimperg replied:
Fixed.

Vierkantor commented on src/data/list/indexes.lean:
```diff
@@ -0,0 +1,163 @@
+  mmap_with_index f as = mmap (uncurry f) (enum as) :=
+by simp only
+    [ mmap_with_index, mmap_with_index_aux_spec,
+      mmap_with_index_aux_eq_mmap_with_index_aux_spec, enum ]
```
```suggestion
    [mmap_with_index, mmap_with_index_aux_spec,
     mmap_with_index_aux_eq_mmap_with_index_aux_spec, enum ]
```

Vierkantor reviewed (changes requested):
Some small formatting suggestions.

JLimperg committed:
Apply formatting suggestions

Co-authored-by: Vierkantor <Vierkantor@users.noreply.github.com>

JLimperg committed:
Generalise mmap_with_index_aux_spec

Vierkantor reviewed (approved):
LGTM.

bors r+

JLimperg committed:
Merge remote-tracking branch 'origin/master' into fold_with_index

JLimperg committed:
Remove unused instance argument

Linter being useful!
